jmespath-core/src/main/java/io/burt/jmespath/function/ArgumentConstraints.java

```diff
@@ -103,195 +103,184 @@ public static ArgumentConstraint expression() {
 
   private ArgumentConstraints() {}
 
-  @SuppressWarnings("serial")
-  static class InternalArgumentTypeException extends FunctionCallException {
-    private final String expectedType;
-    private final String actualType;
+  private static abstract class BaseArgumentConstraint implements ArgumentConstraint {
+    private final int minArity;
+    private final int maxArity;
+    private final String expectedTypeDescription;
 
-    public InternalArgumentTypeException(String expectedType, String actualType) {
-      this(expectedType, actualType, null);
+    public BaseArgumentConstraint(int minArity, int maxArity, String expectedTypeDescription) {
+      this.minArity = minArity;
+      this.maxArity = maxArity;
+      this.expectedTypeDescription = expectedTypeDescription;
     }
 
-    public InternalArgumentTypeException(String expectedType, String actualType, Throwable cause) {
-      super("", cause);
-      this.expectedType = expectedType;
-      this.actualType = actualType;
+    protected <T> ArgumentError checkNoRemaingArguments(Iterator<FunctionArgument<T>> arguments, boolean expectNoRemainingArguments) {
+      if (expectNoRemainingArguments && arguments.hasNext()) {
+        return ArgumentError.createArityError();
+      } else {
+        return null;
+      }
     }
 
-    public String expectedType() { return expectedType; }
+    @Override
+    public int minArity() {
```

Comment:
There's a lesser known (or perhaps just less common) style of writing Java that skips the "get" prefix on getters when there are no setters. Since I more or less never write any setters ever I use that style most of the time. It's maybe coloured a bit by Ruby, too.